### PR TITLE
Extract service implementation

### PR DESCRIPTION
**Pull request by vyzo**
Extract the service implementation from the package, so that there is no dependency on go-libp2p; this allows us to import the package in go-libp2p itself.
The service implementation will live in https://github.com/libp2p/go-libp2p-autonat-svc.
Also extracts IsPublicAddr to go-multiaddr-net.

### CODE REVIEW

**Comment by Stebalien** [comment]
I'm going to apply my "someone else reviews this first" OKR (unless this is *really* burning).

**Comment by vyzo** [comment]
we will need it to merge autorelay.

File: autonat.go [hunk 1]
```diff
@@ -61,12 +63,18 @@ type AmbientAutoNAT struct {
 }
 
 // NewAutoNAT creates a new ambient NAT autodiscovery instance attached to a host
-func NewAutoNAT(ctx context.Context, h host.Host) AutoNAT {
+func NewAutoNAT(ctx context.Context, h host.Host, ga ...GetAddrs) AutoNAT {
+	getAddrs := h.Addrs
+	if len(ga) > 0 {
```
Comment (magik6k) [inline autonat.go]: I'd just make this a static arg and check if it's nil

**Review by magik6k — COMMENTED** [review]
Few nits

File: autonat.go (hunk 1 above)
Comment (vyzo) [reply autonat.go]: ok

Comment (raulk) [inline autonat.go]: Is there a reason we decided against the customary `Option` approach? I prefer that approach, as this appears very limiting in the future.

Comment (vyzo) [reply autonat.go]: done.

File: client.go [hunk 2]
```diff
@@ -27,13 +27,21 @@ type AutoNATError struct {
...
+// GetAddrs is a function that returns the addresses to dial back
+type GetAddrs func() []ma.Multiaddr
+
 // NewAutoNATClient creates a fresh instance of an AutoNATClient
-func NewAutoNATClient(h host.Host) AutoNATClient {
-	return &client{h: h}
+func NewAutoNATClient(h host.Host, ga ...GetAddrs) AutoNATClient {
+	getAddrs := h.Addrs
+	if len(ga) > 0 {
+		getAddrs = ga[0]
```
Comment (raulk) [inline client.go]: Could we get test coverage for this?

File: autonat_test.go [hunk 3]
```diff
@@ -17,24 +23,76 @@ func init() {
...
+	h := bhost.NewBlankHost(swarmt.GenSwarm(t, ctx))
 	a := NewAutoNAT(ctx, h)
+	a.(*AmbientAutoNAT).peers[ash.ID()] = struct{}{}
 
 	return h, a
 }
 
-// Note: these tests assume the host has only private inet addresses!
+func connect(t *testing.T, a, b host.Host) {
+	pinfo := pstore.PeerInfo{ID: a.ID(), Addrs: a.Addrs()}
```
[the remark this answers is not in the document]
Comment (vyzo) [reply autonat_test.go]: yeah, this gets to me every single time as peerstore is a pretty big dependency.
Maybe we should have it in `go-libp2p-peer`?

File: autonat.go (hunk 1 above)
Comment (raulk) [reply autonat.go]: Why not a variadic `Option`? @magik6k @vyzo

Comment (vyzo) [reply autonat.go]: turned into static argument per Magik's suggestion.

File: client.go (hunk 2 above)
Comment (vyzo) [reply client.go]: it's covered in https://github.com/libp2p/go-libp2p-autonat-svc/pull/1 which has an actual service implementation.
It's also implicitly covered in `autonat_test` which uses it internally.

File: autonat_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (raulk) [reply autonat_test.go]: Yeah. I think `host` is a better place as `peer` doesn't depend on multiaddrs. Noted for the (hopefully near) future.

**Review by magik6k — APPROVED** [review]

**Review by raulk — APPROVED** [review]

**Pull request merged**